src/metabase/models/dashboard.clj

```diff
@@ -183,36 +183,62 @@
   (-> dashboard
       (select-keys [:collection_id :description :name :cache_ttl :auto_apply_filters])
       (assoc :cards (vec (for [dashboard-card (ordered-cards dashboard)]
-                           (-> (select-keys dashboard-card [:size_x :size_y :row :col :id :card_id])
-                               (assoc :series (mapv :id (dashboard-card/series dashboard-card)))))))))
+                           (-> (select-keys dashboard-card [:dashboard_tab_id :size_x :size_y :row :col :id :card_id])
+                               (assoc :series (mapv :id (dashboard-card/series dashboard-card)))))))
+      (assoc :tabs (map #(select-keys % [:id :name :position :dashboard_id]) (ordered-tabs dashboard)))))
+
+(defn- revert-tabs
+  [dashboard-id serialized-tabs]
+  (let [current-tabs                       (t2/select [:model/DashboardTab :id :name :position]
+                                                      :dashboard_id dashboard-id)
+        current-tab-ids                    (set (map :id current-tabs))
+        serialized-tab-ids                 (set (map :id serialized-tabs))
+        [delete-ids create-ids update-ids] (diff current-tab-ids serialized-tab-ids)
+        _                                  (when (seq delete-ids) (dashboard-tab/delete-tabs! (seq delete-ids)))
+        to-create-tabs                     (filter #(contains? create-ids (:id %)) serialized-tabs)
+        new-tab-ids                        (when (seq to-create-tabs)
+                                             (t2/insert-returning-pks! :model/DashboardTab to-create-tabs))
+        id->serialized-tab                 (zipmap (map :id serialized-tabs) serialized-tabs)
+        id->current-tab                    (zipmap (map :id current-tabs) current-tabs)]
+       (when (seq update-ids)
+         (dashboard-tab/update-tabs! (map id->current-tab update-ids) (map id->serialized-tab update-ids)))
+   {:old-id->new-id  (zipmap (map :id to-create-tabs) new-tab-ids)}))
+
+(defn- revert-dashcards
```

Comment:
Also realised this looks very similar to `metabase.api.dashboard/do-update-dashcards!`. There's opportunity for reuse here.

Reply:
initially I wanted to do that, but the `create-dashcards!` or `update-dashcards` need to do [param checking](https://github.com/metabase/metabase/blob/b64cd2674581716b5ce68186fee62c2c46f47b83/src/metabase/api/dashboard.clj#L603) that calls to too many fns in the API ns that I can't move it out.